client/state/selectors/is-recurring-payments-stats-nudge-dismissed.js

```diff
@@ -0,0 +1,48 @@
+/**
+ * External dependencies
+ */
+import { last } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { getPreference } from 'state/preferences/selectors';
+
+const WEEK_IN_MS = 7 * 24 * 60 * 60 * 1000;
+
+/**
+ * Returns the last time the nudge was dismissed by the current user or 0 if it was never dismissed
+ *
+ * @param  {object}  state  Global state tree
+ * @param  {number}  siteId The Id of the site
+ * @returns {number}  Timestamp marking the last time the nudge was dismissed
+ */
+const getLastDismissTime = ( state, siteId ) => {
+	const preference = getPreference( state, 'recurring-payments-dismissible-nudge' ) || {};
+	const sitePreference = preference[ siteId ] || [];
+	const lastEvent = last( sitePreference.filter( event => 'dismiss' === event.type ) );
+
+	return lastEvent ? lastEvent.dismissedAt : 0;
+};
+
+/**
+ * Returns true if the Upwork nudge has recently been dismissed by the current user
+ * and this action is still effective for this site
+ *
+ * The conditions for it to be effective (and thus make the nudge invisible) are the following:
+ * - The last time it was dismissed must be less than 2 weeks ago
+ *
+ * @param  {object}  state  Global state tree
+ * @param  {number}  siteId The Id of the site
+ * @returns {boolean} True if the nudge has been dismissed
+ */
+export default function isRecurringPaymentsStatsNudgeDismissed( state, siteId ) {
+	const lastDismissTime = getLastDismissTime( state, siteId );
+
+	// Return false if it has never been dismissed
+	if ( lastDismissTime === 0 ) {
+		return false;
+	}
+
+	return lastDismissTime > Date.now() - 2 * WEEK_IN_MS;
```

Comment:
Just curious; is there a precedent for date math being used in a selector? It makes it non-deterministric which could make it harder to test anything using this selector in the future. Could we just make the second part a helper function somewhere that accepts the date and does the comparison? Then it would at least not be tied to the state.

Reply:
We have the same pattern in:
https://github.com/Automattic/wp-calypso/blob/f2c9d99b07e14ea938829e9b2e5b31b6fb388a90/client/state/selectors/is-gsuite-stats-nudge-dismissed.js#L47
https://github.com/Automattic/wp-calypso/blob/f2c9d99b07e14ea938829e9b2e5b31b6fb388a90/client/state/selectors/is-google-my-business-stats-nudge-dismissed.js#L51

And other selectors for these nudges

Reply:
Thanks for the references. I still think it defeats the purpose of a selector, but ok.